### ISSUE TITLE
Add ad-hoc component definitions to `Checkmatefile`

### ISSUE THREAD

**Issue by larsbutler** [checkmate/chessboard#9]
We need to be able to define custom components in a `Checkmatefile`. After some "whiteboarding" in an etherpad with @ziadsawalha, we think the structure should look  something like the following. It still needs work.

```
blueprint:
  ...  # `blueprint` is already part of Checkmatefiles, and is just shown here for context.
components:
  <name-of-component>:
    commands:
      # these are all lists of shell commands, basically
      start: ["service foo start --port {{ port }}"]
      install: [
        "apt-get update",
        "apt-get install foo",
      ]
      stop: ["service foo stop"]
      kill: ["pkill foo", "rm /tmp/something/to/clean/up"]
      test: ["curl localhost:{{ port }}"]
    options:
      port:
        get_from:
          # env, etcd, and keychain are well-known keywords here
          # try to get the value from these places, in this particular order
          - env: FOO_PORT
          - etcd: /var/foo_port
          - keychain: foo_port
        # if env/etcd/keychain don't have the value we want, use the default:
        default: {{ get_random_free_port() }}
      provides:
        - http
      requires:
        - host: linux
```

Each provider will inherently include a basic set of "primitive" components. For example, a Rackspace provider will include components like load balancers, compute nodes, object storage, and cloud databases (Redis, MongoDB, etc.).

The idea is that custom components can be defined in a Checkmatefile, which are then added to a provider's catalog at runtime.

**Comment by ziadsawalha** [comment]
To guide this discussion we need to think about and precisely word the principle we will use to distinguish between these software component definitions and everything else (Dockerfile, chef, ansible).

We don’t want to re-invent the wheel or attempt to become a software configuration tool. In fact, we want to support and leverage other tools as much as possible.

Our focus is on the basics of the application. And with our programmer hat on, the key pieces of information we want to convey are:
- **what** configuration settings (a.k.a. options, arguments, settings, etc...) can be set on our app
- **how** to set them (config file, command-line argument, keychain, environment variable, etc...)
- **metadata** about them such as data type (strings, integers, comma-separated-strings), do they have constraints (regex, list of acceptable values), sensitivity (are the passwords or private keys), and such...
- **relationships** between various options (ex. are any of them mutually exclusive) and other related services (does the `mongodb_url` argument get populated from the information coming from `require: [mongo]` related service)

This is all data that is relevant to the application and independent of any software configuration tools. In fact, much of this data could be used to generate software configuration artifacts (chef cookbooks, Dockerfile, etc...)

**Comment by larsbutler** [comment]
> To guide this discussion we need to think about and precisely word the principle we will use to distinguish between these software component definitions and everything else (Dockerfile, chef, ansible).

I've amended the issue body to include some provisions for config management providers. I still strongly protest against putting Docker in the same category with chef and ansible. :) One could argue that Docker does config management via `RUN` commands (shell commands, essentially), but that's not unique to Docker. Also, you can use Chef, for example, to [manage and configure Docker containers](https://www.chef.io/solutions/containers/).

> We don’t want to re-invent the wheel or attempt to become a software configuration tool. In fact, we want to support and leverage other tools as much as possible.

Totally agree.

> Our focus is on the basics of the application. And with our programmer hat on, the key pieces of information we want to convey are:

> what configuration settings (a.k.a. options, arguments, settings, etc...) can be set on our app
> how to set them (config file, command-line argument, keychain, environment variable, etc...)
> metadata about them such as data type (strings, integers, comma-separated-strings), do they have constraints (regex, list of acceptable values), sensitivity (are the passwords or private keys), and such...

I'm not so sure we need metadata about the environment variables. Env vars are just strings; let the app which uses them figure out what to do with them. The app should do this anyway (e.g., to ensure that someone doesn't try to use "starboard" for a port value). Perhaps this is something we can add in the future, but for now I would argue, "YAGNI".

> relationships between various options (ex. are any of them mutually exclusive) and other related services (does the mongodb_url argument get populated from the information coming from require: [mongo] related service)

I think I see what you're getting at, but I can't relate it to a practical example. Can you please give one?

> This is all data that is relevant to the application and independent of any software configuration tools. In fact, much of this data could be used to generate software configuration artifacts (chef cookbooks, Dockerfile, etc...)

**Comment by ziadsawalha** [comment]
The distinction between infrastructure and config providers is implementation concerns leaking up, IMO. Our main concern is defining the application, so all we need to say is "I want a MySQL database". Take, for example, RDS or Rackspace Cloud Databases. Are they infrastructure or config management? It doesn't matter and I shouldn't have to remember that RDS is infrastructure and chef configuring MySQL on an EC2 instance is a config manager. They both do the same thing for my app; they give me my MySQL database.

**Comment by ziadsawalha** [comment]
WRT to metadata, our work in this in the past two years is relatively mature. The metadata is extremely helpful for clients and the user experience. We have a lot of up/front, inline validation we can do in he UI based on that metadata. That work we should just port from checkmate/rook (which has already informed OpenStack Heat's HOT syntax and the Rackspace Contril Panel UI/UX) ).
Not all config values are env_vars. ini, YAML, and config stores distinguish data types and using just strings for everything would be throwing away a ton of ready-to-use UI and engine code.

**Comment by ziadsawalha** [comment]
WRT relationships, think about defining checkmate today, it `requires: [{database: mongo}]` and the port and host info from mongo (whether it comes from ObjectRocket or Docker) need to find their way into the CHECKMATE_MONGO_URL env var, the --mongo-url argument, or the keychain entry.

**Comment by ryandub** [comment]
This looks very clean and seems to help with the confusion between Checkmatefiles and Chefmaps.

We should acknowledge upfront that the install, start, stop syntax is very similar to JuJu's approach (https://github.com/charms/kibana/tree/master/hooks). Are there lessons learned from that project that we should adopt? Their concept of hooks seems to extend this past basic functionality. How do we plan to deal with that. If you components section is shell scripts, how do you make your component platform agnostic? If you want to do that with config management, how do you make your start/stop/kill (this seems like a different kind of stop) sections platform agnostic? From your example, providers are only in the install section.

Systemd also has adopted a similar strategy for service files with StartPre, Start StopPre, Stop, etc. (https://coreos.com/docs/launching-containers/launching/getting-started-with-systemd/). Is it beneficial for us to standardize on Systemd and make Systemd templates instead? Systemd is already being used by Debian Jessie, Cent/RHEL 7, Fedora, CoreOS, and soon Ubuntu.

With the Chef provider, what is an example of your `my_cookbook_file`? Is that a Berksfile? Is this being processed by the engine or by the client (application server).

**Comment by larsbutler** [comment]
@ryandub 

> This looks very clean and seems to help with the confusion between Checkmatefiles and Chefmaps.

Cool!

> We should acknowledge upfront that the install, start, stop syntax is very similar to JuJu's approach (https://github.com/charms/kibana/tree/master/hooks). Are there lessons learned from that project that we should adopt? Their concept of hooks seems to extend this past basic functionality. How do we plan to deal with that. 

I don't have a good answer for that; I'd have to dig into JuJu a bit more. Ryan, have you used JuJu before? How do they extend past basic functionality?

> If you components section is shell scripts, how do you make your component platform agnostic? If you want to do that with config management, how do you make your start/stop/kill (this seems like a different kind of stop) sections platform agnostic? From your example, providers are only in the install section.

Simple things first: `stop` is basically a `kill` (graceful shutdown; let it finish what it's doing), `kill` is a `kill -9` (nuke it from orbit). That's the idea.

By "platform agnostic", do you mean "OS agnostic"? If yes, I would say that I wouldn't attempt to make configuration via bash work on all linux flavors, for example. Maybe the story is different for Chef or Ansible, but I don't have much experience with those, so I don't know. What's missing from this component schema is the choice of OS, for which I've added to the checklist in the issue description. I think we need to force the blueprint writer to choose a deployment OS/distro; I don't think it's realistic or feasible to try to make something work on any *nix.

Are we talking about the same problem, or did I misunderstand you?

> Systemd also has adopted a similar strategy for service files with StartPre, Start, StopPre, Stop, etc. (https://coreos.com/docs/launching-containers/launching/getting-started-with-systemd/). Is it beneficial for us to standardize on Systemd and make Systemd templates instead? Systemd is already being used by Debian Jessie, Cent/RHEL 7, Fedora, CoreOS, and soon Ubuntu.

Good point; we could do this. I see the utility of have pre/post start/stop. My inclination, however, is to start with just the basic hooks (install, start, stop, kill) and then add others (prestart, poststart, prestop, poststop, etc.) if and when we need them, because YAGNI. What do you think?

> With the Chef provider, what is an example of your my_cookbook_file? Is that a Berksfile? Is this being processed by the engine or by the client (application server).

Disclaimer: I have zero experience with Chef, so my grasp of Chef terminology, artifacts, and the whole ecosystem is not great. Maybe you can school me a little bit on there. =)

Basically, yes, you specify a Berksfile and it is run on the VM/container/application server _after_ `chessboard`/`checkmate` provider code has actually created the resource. For example, deploying on Rackspace, the provider code would provision a VM, and then run the Berksfile/shell script/whatever to configure the machine. Same general idea with docker and vagrant providers. The shell script/Berksfile wouldn't be run inside the context of the engine.

**Comment by larsbutler** [comment]
After working on the docker provider a bit, I think it make sense to make the `components` section a `list` instead of a `dict`/mapping, don't put the name in mapping key, and instead, put the name as a required attribute of the component. This relieves a lot of headaches in handling these structures in the provider code.

**Comment by ziadsawalha** [comment]
We've historically chosen to optimize for the author/reader of the blueprint. And with that we accept headaches in the code since those headaches are written once and can be constrained by tests. I.e. headaches in the code should not determine the syntax of the blueprint.
Can you give an example of the headaches in the code?

**Comment by larsbutler** [comment]
If you want to load a component just from a file (component.yaml), and `name` is not required, where do you get the component name? Probably it would be okay to keep the current mapping/dict structure (keyed by component name), but the `name` attribute should be required as well.

I'll try to work around it, but I just wanted to flag this as an issue in writing code to deal with this format. Obviously we wouldn't anticipate this, because this is the first time we're writing code based on the format.
